src/libflake/flake/flake.cc

```diff
@@ -823,20 +826,63 @@ void initLib(const Settings & settings)
 {
     auto prim_getFlake = [&settings](EvalState & state, const PosIdx pos, Value * * args, Value & v)
     {
-        std::string flakeRefS(state.forceStringNoCtx(*args[0], pos, "while evaluating the argument passed to builtins.getFlake"));
-        auto flakeRef = parseFlakeRef(state.fetchSettings, flakeRefS, {}, true);
-        if (state.settings.pureEval && !flakeRef.input.isLocked())
-            throw Error("cannot call 'getFlake' on unlocked flake reference '%s', at %s (use --impure to override)", flakeRefS, state.positions[pos]);
-
-        callFlake(state,
-            lockFlake(settings, state, flakeRef,
-                LockFlags {
-                    .updateLockFile = false,
-                    .writeLockFile = false,
-                    .useRegistries = !state.settings.pureEval && settings.useRegistries,
-                    .allowUnlocked = !state.settings.pureEval,
-                }),
-            v);
+        state.forceValue(*args[0], pos);
+
+        LockFlags lockFlags {
+            .updateLockFile = false,
+            .writeLockFile = false,
+            .warnModifiedLockFile = false,
+            .useRegistries = !state.settings.pureEval && settings.useRegistries,
+            .allowUnlocked = !state.settings.pureEval,
+        };
+
+        auto flakeRef =
+            args[0]->type() == nString
+            ? ({
+                std::string flakeRefS(state.forceStringNoCtx(*args[0], pos, "while evaluating the argument passed to builtins.getFlake"));
+                auto flakeRef = parseFlakeRef(state.fetchSettings, flakeRefS, {}, true);
+                if (state.settings.pureEval && !flakeRef.input.isLocked())
+                    throw Error("cannot call 'getFlake' on unlocked flake reference '%s', at %s (use --impure to override)", flakeRefS, state.positions[pos]);
+                flakeRef;
+            })
+            : ({
+                auto flakeInput = parseFlakeInput(state, std::nullopt, args[0], pos, {}, {});
+
+                /* Convert the result of parseFlakeInput() into a
+                   overrides map and a top-level flakeref. */
+                std::function<void(const InputPath & inputPath, const FlakeInput & input)> recurse;
+
+                recurse = [&](const InputPath & inputPath, const FlakeInput & input)
+                {
+                    if (!input.ref)
```

Comment:
It's not strictly necessary to use `std::function` and its inherent indirection for recursive lambdas. Though in this case this shouldn't matter much. With C++23 it's possible to use "deducing this" https://en.cppreference.com/w/cpp/language/member_functions#Explicit_object_member_functions. For pre C++23 one can use the following hack:

```cpp
auto recurse_impl = [&](auto& self, const InputPath & inputPath, const FlakeInput & input) {
....
  recurse_impl(self, ....);
};

auto recurse = [&](const InputPath & inputPath, const FlakeInput & input) {
  recurse_impl(recurse_impl, inputPath, input);
};
```

This avoid the virtual call via `std::function`

Reply:
Wow nice, didn't know about "deducing this". Apparently it will be avalable in gcc 14.